File: apk_monytix/app/src/main/java/com/example/apk/console/tabs/SpendingTabView.kt
package com.example.apk.console.tabs

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.example.apk.console.MolyConsoleViewModel
import com.example.apk.ui.components.GlassCard
import com.example.apk.ui.theme.Gold
import com.example.apk.ui.theme.TextPrimary
import com.example.apk.ui.theme.TextSecondary
import com.example.apk.console.utils.formatCurrency
import java.util.Locale

@Composable
fun SpendingTabView(viewModel: MolyConsoleViewModel) {
    val monthlySpending by viewModel.monthlySpending.collectAsStateWithLifecycle()
    val spendingByCategory by viewModel.spendingByCategory.collectAsStateWithLifecycle()
    val isLoading by viewModel.isSpendingLoading.collectAsStateWithLifecycle()
    val error by viewModel.spendingError.collectAsStateWithLifecycle()
    
    when {
        isLoading -> {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(color = Gold)
            }
        }
        error != null -> {
            SpendingErrorState(error = error ?: "Unknown error") {
                viewModel.loadSpending()
            }
        }
        spendingByCategory.isEmpty() -> {
            SpendingEmptyState()
        }
        else -> {
            SpendingContent(
                monthlySpending = monthlySpending,
                spendingByCategory = spendingByCategory
            )
        }
    }
}

@Composable
fun SpendingContent(
    monthlySpending: Double,
    spendingByCategory: List<com.example.apk.console.models.CategorySpending>
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(20.dp)
    ) {
        // This Month's Spending
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "This Month's Spending",
                fontSize = 16.sp,
                fontWeight = FontWeight.Medium,
                color = TextSecondary
            )
            
            Text(
                text = formatCurrency(monthlySpending),
                fontSize = 32.sp,
                fontWeight = FontWeight.Bold,
                color = Color(0xFFE53935)
            )
        }
        
        // Spending by Category
        Text(
            text = "Spending by Category",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            color = TextPrimary,
            modifier = Modifier.fillMaxWidth()
        )
        
        spendingByCategory.forEach { category ->
            CategorySpendingCard(
                category = category,
                totalSpending = monthlySpending
            )
        }
    }
}

@Composable
fun CategorySpendingCard(
    category: com.example.apk.console.models.CategorySpending,
    totalSpending: Double
) {
    GlassCard(
        padding = PaddingValues(16.dp),
        cornerRadius = 16.dp
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = category.category,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = TextPrimary
                )
                
                Text(
                    text = formatCurrency(category.amount),
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    color = TextPrimary
                )
            }
            
            // Progress Bar
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                LinearProgressIndicator(
                    progress = { (category.percentage / 100).toFloat() },
                    modifier = Modifier
                        .weight(1f)
                        .height(8.dp),
                    color = Gold,
                    trackColor = TextSecondary.copy(alpha = 0.2f)
                )
                
                Text(
                    text = "${String.format(Locale.getDefault(), "%.1f", category.percentage)}%",
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium,
                    color = TextSecondary,
                    modifier = Modifier.width(50.dp)
                )
            }
            
            Text(
                text = "${category.transactionCount} transactions",
                fontSize = 12.sp,
                color = TextSecondary.copy(alpha = 0.6f)
            )
        }
    }
}

@Composable
private fun SpendingErrorState(error: String, onRetry: () -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(60.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        Icon(
            imageVector = Icons.Default.Error,
            contentDescription = null,
            tint = Color(0xFFE53935).copy(alpha = 0.7f),
            modifier = Modifier.size(48.dp)
        )
        
        Text(
            text = "Unable to Load Spending",
            fontSize = 18.sp,
            fontWeight = FontWeight.Bold,
            color = TextPrimary
        )
        
        Text(
            text = error,
            fontSize = 14.sp,
            color = TextSecondary,
            modifier = Modifier.padding(horizontal = 32.dp)
        )
        
        Button(
            onClick = onRetry,
            colors = ButtonDefaults.buttonColors(containerColor = Gold)
        ) {
            Text("Retry", color = com.example.apk.ui.theme.Charcoal)
        }
    }
}

@Composable
private fun SpendingEmptyState() {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(60.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        Icon(
            imageVector = Icons.Default.AttachMoney,
            contentDescription = null,
            tint = TextSecondary.copy(alpha = 0.5f),
            modifier = Modifier.size(48.dp)
        )
        
        Text(
            text = "No Spending Data",
            fontSize = 18.sp,
            fontWeight = FontWeight.SemiBold,
            color = TextPrimary
        )
        
        Text(
            text = "Upload statements to see your spending breakdown",
            fontSize = 14.sp,
            color = TextSecondary,
            modifier = Modifier.padding(horizontal = 32.dp)
        )
    }
}

File: ios_monytix/ios_monytix/console/SpendingTabView.swift
//
//  SpendingTabView.swift
//  ios_monytix
//
//  Created by santosh on 05/01/26.
//

import SwiftUI

struct SpendingTabView: View {
    @ObservedObject var viewModel: MolyConsoleViewModel
    
    private let goldColor = Color(red: 0.831, green: 0.686, blue: 0.216)
    
    var body: some View {
        ScrollView {
            VStack(spacing: 20) {
                if viewModel.isSpendingLoading {
                    ProgressView()
                        .tint(goldColor)
                        .padding(.top, 40)
                } else if let error = viewModel.spendingError {
                    errorState(error)
                } else {
                    spendingContent
                }
            }
            .padding(20)
        }
        .refreshable {
            await viewModel.loadSpending()
        }
    }
    
    private var spendingContent: some View {
        VStack(alignment: .leading, spacing: 20) {
            // This Month's Spending
            VStack(alignment: .leading, spacing: 8) {
                Text("This Month's Spending")
                    .font(.system(size: 16, weight: .medium))
                    .foregroundColor(.gray.opacity(0.7))
                
                Text(formatCurrency(viewModel.monthlySpending))
                    .font(.system(size: 32, weight: .bold))
                    .foregroundColor(.red)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.bottom, 8)
            
            // Spending by Category
            if !viewModel.spendingByCategory.isEmpty {
                Text("Spending by Category")
                    .font(.system(size: 20, weight: .bold))
                    .foregroundColor(.white)
                    .frame(maxWidth: .infinity, alignment: .leading)
                
                ForEach(viewModel.spendingByCategory) { category in
                    CategorySpendingCard(category: category, totalSpending: viewModel.monthlySpending)
                }
            } else {
                emptyState
            }
        }
    }
    
    private func errorState(_ error: String) -> some View {
        VStack(spacing: 16) {
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.system(size: 48))
                .foregroundColor(.red.opacity(0.7))
            
            Text("Unable to Load Spending")
                .font(.system(size: 18, weight: .bold))
                .foregroundColor(.white)
            
            Text(error)
                .font(.system(size: 14))
                .foregroundColor(.gray.opacity(0.7))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)
            
            Button("Retry") {
                Task {
                    await viewModel.loadSpending()
                }
            }
            .buttonStyle(GoldButtonStyle())
            .padding(.top, 8)
        }
        .padding(.top, 60)
    }
    
    private var emptyState: some View {
        VStack(spacing: 16) {
            Image(systemName: "dollarsign.circle.fill")
                .font(.system(size: 48))
                .foregroundColor(.gray.opacity(0.5))
            
            Text("No Spending Data")
                .font(.system(size: 18, weight: .semibold))
                .foregroundColor(.white)
            
            Text("Upload statements to see your spending breakdown")
                .font(.system(size: 14))
                .foregroundColor(.gray.opacity(0.7))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)
        }
        .padding(.top, 40)
    }
    
    private func formatCurrency(_ amount: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencySymbol = "₹"
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: amount)) ?? "₹0"
    }
}

// MARK: - Category Spending Card

struct CategorySpendingCard: View {
    let category: CategorySpending
    let totalSpending: Double
    
    private let goldColor = Color(red: 0.831, green: 0.686, blue: 0.216)
    
    var body: some View {
        GlassCard(padding: 16, cornerRadius: 16) {
            VStack(alignment: .leading, spacing: 12) {
                HStack {
                    Text(category.category)
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(.white)
                    
                    Spacer()
                    
                    Text(formatCurrency(category.amount))
                        .font(.system(size: 16, weight: .bold))
                        .foregroundColor(.white)
                }
                
                HStack(spacing: 8) {
                    // Progress Bar
                    GeometryReader { geometry in
                        ZStack(alignment: .leading) {
                            RoundedRectangle(cornerRadius: 4)
                                .fill(Color.gray.opacity(0.2))
                                .frame(height: 8)
                            
                            RoundedRectangle(cornerRadius: 4)
                                .fill(goldColor)
                                .frame(width: geometry.size.width * (category.percentage / 100), height: 8)
                        }
                    }
                    .frame(height: 8)
                    
                    Text("\(String(format: "%.1f", category.percentage))%")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundColor(.gray.opacity(0.7))
                        .frame(width: 50, alignment: .trailing)
                }
                
                Text("\(category.transactionCount) transactions")
                    .font(.system(size: 12))
                    .foregroundColor(.gray.opacity(0.6))
            }
        }
    }
    
    private func formatCurrency(_ amount: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencySymbol = "₹"
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: amount)) ?? "₹0"
    }
}

#Preview {
    let authService = AuthService()
    return SpendingTabView(viewModel: MolyConsoleViewModel(authService: authService))
        .background(Color(red: 0.18, green: 0.18, blue: 0.18))
}

